lib/optimize/CommonsChunkPlugin.js

```diff
@@ -24,158 +24,346 @@ The available options are:
+		const asyncChunk = compilation.addChunk(typeof asyncOption === "string" ? asyncOption : undefined);
+		asyncChunk.chunkReason = "async commons chunk";
+		asyncChunk.extraAsync = true;
+		asyncChunk.addParent(targetChunk);
+		return asyncChunk;
+	}
+
+	// If minChunks is a function use that
+	// otherwhise check if a module is used at least minChunks or 2 or usedChunks.length time
+	getModuleFilter(minChunks, targetChunk, usedChunksLength) {
+		if(typeof minChunks === "function") {
+			return minChunks;
+		}
+		const minCount = (minChunks || Math.max(2, usedChunksLength));
+		const isUsedAtLeastMinTimes = (module, count) => count >= minCount;
+		return isUsedAtLeastMinTimes;
+	}
+
+	getExtractableModules(minChunks, usedChunks, targetChunk) {
+		if(minChunks === Infinity) {
+			return [];
+		}
+
+		// count how many chunks contain a module
+		const commonModulesToCountMap = usedChunks.reduce((map, chunk) => {
+			for(let module of chunk.modules) {
+				let count = map.has(module) ? map.get(module) : 0;
+				map.set(module, count + 1);
+			}
+			return map;
+		}, new Map());
+
+		// filter by minChunks
+		const moduleFilterCount = this.getModuleFilter(minChunks, targetChunk, usedChunks.length);
+		// filter by condition
+		const moduleFilterCondition = (module, chunk) => {
+			if(!module.chunkCondition) {
+				return true;
+			}
+			return module.chunkCondition(chunk);
+		};
+
+		return Array.from(commonModulesToCountMap).filter(entry => {
+			const module = entry[0];
+			const count = entry[1];
+			// if the module passes both filters, keep it.
+			return moduleFilterCount(module, count) && moduleFilterCondition(module, targetChunk);
+		}).map(entry => entry[0]);
+	}
+
+	calculateModuleSize(modules) {
+		return modules.reduce((count, module) => count + module.size(), 0);
+	}
+
+	extractModulesAndReturnAffectedChunks(reallyUsedModules, usedChunks) {
+		return reallyUsedModules.reduce((affectedChunksSet, module) => {
+			for(let chunk of usedChunks) {
+				// removeChunk returns true if the chunk was contained and succesfully removed
+				// false if the module did not have a connection to the chunk in question
+				if(module.removeChunk(chunk)) {
+					affectedChunksSet.add(chunk);
+				}
+			}
+			return affectedChunksSet;
+		}, new Set());
+	}
+
+	addExtractedModulesToTargetChunk(chunk, modules) {
+		for(let module of modules) {
+			chunk.addModule(module);
+			module.addChunk(chunk);
+		}
+	}
+
+	makeTargetChunkParentOfAffectedChunks(usedChunks, commonChunk) {
+		for(let chunk of usedChunks) {
+			// set commonChunk as new sole parent
+			chunk.parents = [commonChunk];
+			// add chunk to commonChunk
+			commonChunk.addChunk(chunk);
+
+			for(let entrypoint of chunk.entrypoints) {
+				entrypoint.insertChunk(commonChunk, chunk);
+			}
+		}
+	}
+
+	moveExtractedChunkBlocksToTargetChunk(chunks, commonChunk) {
+		for(let chunk of chunks) {
+			// only for non initial chunks
+			// TODO: why?
```

Comment:
I was also wondering while reading the old code.

I guess this doesn't make sense and we can remove the check.